File: BaekJoon/Category/Graph/Depth-First-Search/1260_DFS와BFS.cpp
#include <bits/stdc++.h>

using namespace std;

const int MAX = 1001;

int N, M, V; 
bool visited[MAX];
vector<int> graph[MAX];

void dfs(int start) {
    stack<int> s;
    s.push(start);
    
    while (!s.empty()) {
        int x = s.top();
        s.pop();
        if (visited[x]) continue;

        visited[x] = true;
        cout << x << ' ';

        sort(graph[x].begin(), graph[x].end(), greater<int>());
        for (int i : graph[x]) {
            if (!visited[i]) {
                s.push(i);
            }
        }
    }
    cout << '\n';
}

void bfs(int start) {
    queue<int> q;
    q.push(start);

    while (!q.empty()) {
        int x = q.front();
        q.pop();
        if (visited[x]) continue;

        visited[x] = true;
        cout << x << ' ';

        sort(graph[x].begin(), graph[x].end());
        for (int i : graph[x]) {
            if (!visited[i]) {
                q.push(i);
            }
        }
    }
    cout << '\n';
}

int main(void) {
    ios::sync_with_stdio(false);  cin.tie(NULL);  cout.tie(NULL);

    cin >> N >> M >> V;

    for (int i = 0 ; i < M; i++) {
        int start, end;  
        cin >> start >> end;
        
        graph[start].push_back(end);
        graph[end].push_back(start);
    }

    dfs(V);
    memset(visited, 0, 1001);
    bfs(V);

    return 0;
}

File: BaekJoon/Python/1260_DFS와BFS.py
from collections import deque
import sys


def dfs(graph, visited, v):
    # 현재 노드 방문 처리
    visited[v] = True
    print(v, end=' ')

    # 현재 노드와 연결된 다른 노드를 재귀적으로 방문
    for i in graph[v]:
        if not visited[i]:
            dfs(graph, visited, i)


def bfs(graph, visited, start):
    queue = deque([start])

    visited[start] = True

    while queue:
        x = queue.popleft()
        print(x, end=' ')

        for i in graph[x]:
            if not visited[i]:
                queue.append(i)
                visited[i] = True


n, m, v = map(int, input().split())
graph = [[] for _ in range(n + 1)]
for _ in range(m):
    a, b = map(int, sys.stdin.readline().split())
    graph[a].append(b)
    graph[b].append(a)

for i in range(1, n + 1):
    graph[i].sort()

visited = [False] * (n + 1)
dfs(graph, visited, v)
print()
visited = [False] * (n + 1)
bfs(graph, visited, v)
